archspec/cpu/detect.py

```diff
@@ -206,11 +206,26 @@ def host():
     # Get a list of possible candidates for this micro-architecture
     candidates = compatible_microarchitectures(info)
 
+    # Sorting criteria for candidates
+    def sorting_fn(item):
+        return len(item.ancestors), len(item.features)
+
+    # Get the best generic micro-architecture
+    generic_candidates = [c for c in candidates if c.vendor == "generic"]
+    best_generic = sorted(generic_candidates, key=sorting_fn, reverse=True)[0]
+
+    # Filter the candidates to be descendant of the best generic candidate.
+    # This is to avoid that the lack of a niche feature that can be disabled
+    # from e.g. BIOS prevents detection of a reasonably performant architecture
+    candidates = [c for c in candidates if c > best_generic]
+
+    # If we don't have candidates, return the best generic micro-architecture
+    if not candidates:
+        return best_generic
```

Comment:
Can we add support for `host(generic=True)` to force returning a generic result, even if better non-generic compatible candidates are available?

Reply:
I think a better approach, to keep a consistent API, is to have some property on the microarchitecture that gets returned (this would be consistent with the `uarch.family` property)

Reply:
`uarch.level`?

Reply:
If we follow ARM naming it could also be `uarch.revision`

Reply:
x86_64 ones are called levels, which I kind of like.  I thought the arm ones were just called versions.

Reply:
I'm fine with any, even though "version" is a bit generic and incurs the issue of having `config.h` among your application headers :laughing: 

What if we implement one of `revision` or `level` to do the computation and use the other as an alias:
```python
@property
def level(self):
    ...
    return value

@property
def revision(self):
    return self.level
```
?

Reply:
@boegel @tgamblin Let me know what would be the preferred name/names here and I'll push an implementation.

Reply:
Implemented `uarch.level`, let me know if other changes are needed.

Reply:
It's hard to come up with an unambiguous name, but how about `generation`?

`revision` makes sense to me too, `level` a bit less. I agree that `version` is not a good pick.

Reply:
The term "generation" is already taken and is used for Power architectures :grimacing:

Reply:
You mean in `archspec`? In what sense?

Having both `generation` and `revision` or `level` isn't exactly a consistent API either?

Reply:
https://github.com/archspec/archspec/blob/130607c373fd88cd3c43da94c0d3afd3a44084b0/archspec/cpu/microarchitecture.py#L70-L72

Reply:
That seems to fit quite well with what we're discussing here? Other than the `int` type then, perhaps...

Reply:
What about `uarch.generic`?

Reply:
`uarch.generic` is fine with me. @boegel ?

Reply:
Makes total sense to me too :+1:

Reply:
Fixed. Let me know if there are other changes to be done.

Reply:
@boegel